src/compiler/sys.ts

```diff
@@ -293,6 +293,50 @@ namespace ts {
                     removeFile: removeFile
                 };
             }
+            
+            function createWatchedFileSet() {
+                let watchedDirectories: { [path: string]: FileWatcher } = {};
+                let watchedFiles: { [fileName: string]: (fileName: string, removed?: boolean) => void; } = {};
+                
+                function addFile(fileName: string, callback: (fileName: string, removed?: boolean) => void): WatchedFile {
+                    const file: WatchedFile = { fileName, callback };
+                    let watchedPaths = Object.keys(watchedDirectories);
+                    // Try to find parent paths that are already watched. If found, don't add directory watchers
+                    let watchedParentPaths = watchedPaths.filter(path => fileName.indexOf(path) === 0);
+                    // If adding new watchers, try to find children paths that are already watched. If found, close them. 
+                    if (watchedParentPaths.length === 0) {
+                        let pathToWatch = ts.getDirectoryPath(fileName);
+                        for (let watchedPath in watchedDirectories) {
+                            if (watchedPath.indexOf(pathToWatch) === 0) {
+                                watchedDirectories[watchedPath].close();
+                                delete watchedDirectories[watchedPath];
+                            }
+                        }
+                        watchedDirectories[pathToWatch] = _fs.watch(
+                            pathToWatch, 
+                            (eventName: string, relativeFileName: string) => fileEventHandler(eventName, ts.normalizePath(ts.combinePaths(pathToWatch, relativeFileName)))
+                        );
+                    }
+                    watchedFiles[fileName] = callback;
+                    return { fileName, callback }
+                }
+                
+                function removeFile(file: WatchedFile) {
+                    delete watchedFiles[file.fileName];

```

Comment:
but we are still watching the directory anyways? when do a directory watcher ever dies?